worlds/marioland2/logic.py

```diff
@@ -0,0 +1,608 @@
+from .locations import level_name_to_id
+
+
+def is_auto_scroll(state, player, level):
+    level_id = level_name_to_id[level]
+    if state.has_any(["Cancel Auto Scroll", f"Cancel Auto Scroll - {level}"], player):
+        return False
+    return state.multiworld.worlds[player].auto_scroll_levels[level_id] > 0
+
+
+def has_pipe_right(state, player):
+    return state.has_any(["Pipe Traversal - Right", "Pipe Traversal"], player)
+
+
+def has_pipe_left(state, player):
+    return state.has_any(["Pipe Traversal - Left", "Pipe Traversal"], player)
+
+
+def has_pipe_down(state, player):
+    return state.has_any(["Pipe Traversal - Down", "Pipe Traversal"], player)
+
+
+def has_pipe_up(state, player):
+    return state.has_any(["Pipe Traversal - Up", "Pipe Traversal"], player)
+
+
+def has_level_progression(state, item, player, count=1):
+    return state.count(item, player) + (state.count(item + " x2", player) * 2) >= count
+
+
+def mushroom_zone_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Mushroom Zone")
+    reachable_coins = 38
+    if state.has_any(["Mushroom", "Fire Flower"], player) or not auto_scroll:
+        # Was able to get all but 1, being lenient.
+        reachable_coins += 2
+    if has_pipe_down(state, player):
+        # There's 24 in each of the underground sections.
+        # The first one requires missing some question mark blocks if auto scrolling (the last +4).
+        # If you go in the second without pipe up, you can get everything except the last 5 plus the ones in the first
+        # underground section.
+        reachable_coins += 19
+        if has_pipe_up(state, player) or not auto_scroll:
+            reachable_coins += 5
+        if has_pipe_up(state, player):
+            reachable_coins += 20
+            if not auto_scroll:
+                reachable_coins += 4
+    return coins <= reachable_coins
+
+
+def tree_zone_1_normal_exit(state, player):
+    return has_pipe_right(state, player) or state.has("Tree Zone 2 Midway Bell", player),
+
+
+def tree_zone_1_coins(state, player, coins):
+    return coins <= 87 or not is_auto_scroll(state, player, "Tree Zone 1")
+
+
+def tree_zone_2_secret_exit(state, player):
+    return has_pipe_right(state, player) and state.has("Carrot", player)
+
+
+def tree_zone_2_midway_bell(state, player):
+    return has_pipe_right(state, player) or state.has("Tree Zone 2 Midway Bell", player)
+
+
+def tree_zone_2_coins(state, player, coins):
+    auto_scroll = is_auto_scroll(state, player, "Tree Zone 2")
+    reachable_coins = 18
+    if has_pipe_right(state, player):
+        reachable_coins += 38
+        if state.has("Carrot", player):
+            reachable_coins += 12
+            if not auto_scroll:
+                reachable_coins += 30
+    elif state.has("Tree Zone 2 Midway Bell", player):
+        reachable_coins = 30
```

Comment:
This is inconsistent with how almost the entire file does this
```suggestion
        reachable_coins += 12
```